### ISSUE TITLE
Disassembler rounds float constants that are near-to-1 up to 1

### ISSUE THREAD

**Issue by dneto0** [KhronosGroup/SPIRV-Tools#882]
Example, a float value of 0x1.fffffep-1 is very nearly 1 but is not 1.
It should be displayed in a bit-preserving way, but instead is rounded up to 1.

So we get 
  OpConstant %float 1

when we should have had
   OpConstant %float 0x1.fffffep-1

**Comment by dneto0** [comment]
I think people commonly want the default floating point string that their platform C++ library provides.  We should keep that as the default.

So I think we should add a disassembler option to force hex floats, for all float constants in the module. That gives us bit-for-bit preservation for floats through disassembler -> assembler round trip.

**Comment by b-sumner** [comment]
It seems odd to me to make exact preservation of floating point constants through disassemble -> assemble just an option.  I'd say it is a requirement.  I don't really see any harm in outputting 0.1f as 0x1.99999ap-4.

**Comment by dneto0** [comment]
> It seems odd to me to make exact preservation of floating point constants through disassemble -> assemble just an option. I'd say it is a requirement. I don't really see any harm in outputting 0.1f as 0x1.99999ap-4.

Most people aren't wired to read hex floating point very well. It's not the default for any other environment.  I think I'd have a lot of complaints if I flipped the default at this point.

Also, given variability across implementations, it's hard to identify "the" subset of cases that would benefit from bit-to-bit accuracy of hex floats.

**Comment by b-sumner** [comment]
Should the disassembler issue warnings when it outputs floating point values that will not assemble to the same bits?

**Comment by dneto0** [comment]
This appears to be fixed by #1306.

This input (from original report)
```
%float = OpTypeFloat 32
%near1 = OpConstant %float   0x1.fffffep-1
```
Assembles to a binary that disassembles to this:

```
; SPIR-V
; Version: 1.3
; Generator: Khronos SPIR-V Tools Assembler; 0
; Bound: 3
; Schema: 0
      %float = OpTypeFloat 32
%float_0_99999994 = OpConstant %float 0.99999994
```

That then re-assembles to a binary with same binary contents as the original (byte for byte equality).

So I think the problem is resolved.
